lambda-source/index.js

```diff
@@ -62,18 +54,49 @@ exports.handler = async (event, context, callback) => {
           success_location: REDIRECT_URL,
           error_location: REDIRECT_URL
         }
-      })
-    }
+      }
+    )
+    return data;
   } catch (err) {
-    console.error("Error: ", err);
+    console.error("Form Error: ", err);
+    return false;
   }
+}
 
-  let redirectURL = REDIRECT_URL;
-  var response = {
-    "statusCode": 302,
-    "headers": {
-      Location: redirectURL,
-    }
-  };
-  callback(null, response);
-};
\ No newline at end of file
+const Responses = {
+  Unauthorized: () => ({ statusCode: 401 }),
+  BadRequest: () => ({ statusCode: 400 }),
+  Redirect: (Location) => ({ statusCode: 302, headers: { Location }})
+}
+
+async function proxyForm(event, _context) {
+  const buff = Buffer.from(event.body, 'base64');
+  const text = buff.toString('ascii');
+  const recaptchaPair = text.split('&').find(pair => pair.startsWith('g-recaptcha-response'));
+  
+  if (!recaptchaPair) {
+    // reCaptcha not provided at all
+    // maybe bad request?
+    return Errors.Unauthorized()
+  }
+  
+  const [_key, captchaResponse] = recaptchaPair.split("=");
+  console.log("Provided reCaptcha value:", captchaResponse)
+  
+  const isRecaptchaValid = await verifyRecaptcha(captchaResponse);
+  if (!isRecaptchaValid) {
+    return Errors.Unauthorized()
```

Comment:
getting an error for these, not sure if this is a runtime incompatibility or if we need to do some compiling?

> | 1646354789732 | 2022-03-04T00:46:29.732Z c04d44d8-6abd-4c02-ba9b-ad4a4171bfe9 ERROR Invoke Error  {"errorType":"ReferenceError","errorMessage":"Errors is not defined","stack":["ReferenceError: Errors is not defined","    at proxyForm (/var/task/index.js:88:5)","    at processTicksAndRejections (internal/process/task_queues.js:95:5)"]}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |

Reply:
Interesting! I'll double check, assuredly I did something wrong when I was dumping everything together.

Reply:
also, we should ways return a redirect, whether successfully verified or not, so we could do something like

```
const Responses = {
  Redirect: (Location) => ({ statusCode: 302, headers: { Location }})
}

async function proxyForm(event, _context) {
  const buff = Buffer.from(event.body, 'base64');
  const text = buff.toString('ascii');
  const recaptchaPair = text.split('&').find(pair => pair.startsWith('g-recaptcha-response'));
  
  if (!recaptchaPair) {
    // reCaptcha not provided at all
    // maybe bad request?
    console.log("No Captcha provided, aborting and finishing redirect.")
    return Responses.Redirect(REDIRECT_URL);
  }
  
  const [_key, captchaResponse] = recaptchaPair.split("=");
  console.log("Provided reCaptcha value:", captchaResponse)
  
  const isRecaptchaValid = await verifyRecaptcha(captchaResponse);
  if (!isRecaptchaValid) {
    console.log("Captcha failed, aborting and finishing redirect.")
    return Responses.Redirect(REDIRECT_URL);
  }

  const formResponse = await sendForm(text);
  if (!formResponse) {
    console.log("Error submitting to pardot, aborting and finishing redirect.")
    return Responses.Redirect(REDIRECT_URL);
  }

  return Responses.Redirect(REDIRECT_URL);
}

```